### DISCUSSION TITLE
Slots: programmatic usage improvements

### DISCUSSION THREAD

**Discussion by Akryum** [vuejs/rfcs#453] [RFC Discussions]
- Start Date: 2022-06-27
- Target Major Version: 3.x
- Reference Issues: -
- Implementation PR: -

# Summary

- Introduce an helper to make detecting slots with empty content easier
- Allow rendering arrays of vnodes in the template directly
- (?) Allow referencing computed properties as template tags when not using `<script setup>`

# Basic example

```vue
<script setup>
import { computed, useSlots, getFilledSlot } from 'vue'

const slots = useSlots()
// Array of vnodes or null if empty
const FooSlot = computed(() => getFilledSlot(slots?.foo({ someProp: 'hello' })))
</script>

<template>
  <!-- Check if FooSlot is null (empty slot) -->
  <div v-if="FooSlot" class="some-padding">
    <!-- Render array of vnodes -->
    <FooSlot />
  </div>
</template>
```

# Motivation

It's currently cumbersome and non idiomatic to check if a slot is empty to conditionally render parts of the template. For example, consider this modal component:

```vue
<template>
  <modal>
    <div class="modal-header">
      <slot name="header" />
    </div>
    <slot />
  </modal>
</template>
```

We probably don't want to render the `.modal-header` element when the user of the component doesn't put content in the `header` slot.

```vue
<template>
  <modal>
    <div v-if="$slots.header" class="modal-header">
      <slot name="header" />
    </div>
    <slot />
  </modal>
</template>
```

Vue currently only populates the `header` function on `$slots` if it is used in the template.

```vue
<template>
  <MyModal>
    Content
  </MyModal>
</template>
```

```js
// Result:
console.log(this.$slots)
{}
```

```vue
<template>
  <MyModal>
    <template #header>
      Title
    </template>
    Content
  </MyModal>
</template>
```

```js
// Result:
console.log(this.$slots)
{
  header: () => [ /* array of vnodes */ ]
}
```

But this falls apart as soon as you have some conditionals:

```vue
<template>
  <MyModal>
    <template #header>
      <span v-if="someCondition">Title</span>
      <span v-else-if="someOtherCondition">Title 2</span>
    </template>
    Content
  </MyModal>
</template>
```

`$slots.header` is still populated with the slot render function:

```js
// Result:
console.log(this.$slots)
{
  header: () => [ /* array of vnodes */ ]
}
```

So we need to programmatically "render" the slot to check if it's really empty by calling the function:

```js
const slots = useSlots()
const HeaderSlot = computed(() => slots.header?.())
```

This also allows us to check for emptiness in case we pass some props to the slot:

```vue
<template>
  <MyModal>
    <template #header="{ title }">
      <span v-if="title.startsWith('a')">Title</span>
      <span v-else-if="title.startsWith('b')">Title 2</span>
    </template>
    Content
  </MyModal>
</template>
```

```js
// MyModal
const title = ref('a')
const slots = useSlots()
const HeaderSlot = computed(() => slots.header?.({ title: title.value }))
```

We could then check for the length of the vnodes array:

```vue
<template>
  <modal>
    <div v-if="HeaderSlot.length" class="modal-header">
      <slot name="header" title="title" />
    </div>
    <slot />
  </modal>
</template>
```

There are two remaining problems here:
1. The `HeaderSlot` array of vnodes could contain only "invalid" vnodes such as comments or empty fragments.
2. We unnecessarily call the slot function again in the template with `<slot/>` and we need to duplicate the slot props.

Fixing problem 1 requires using an internal function called [`ensureValidVNode`](https://github.com/vuejs/core/blob/25f7a16a6eccbfa8d857977dcf1f23fb36b830b5/packages/runtime-core/src/helpers/renderSlot.ts#L84-L97) to check is the array of vnodes contains other vnodes than comments and empty fragments.

That way we can check if the result is `null`, meaning empty:

```vue
<template>
  <modal>
    <div v-if="HeaderSlot" class="modal-header">
      <slot name="header" title="title" />
    </div>
    <slot />
  </modal>
</template>
```

Fixing problem 2 requires directly using in the template the result we have in the script. The most literal way would be to loop over the vnodes:

```vue
<template>
  <modal>
    <div v-if="HeaderSlot" class="modal-header">
      <component v-for="(s, i) in HeaderSlot" :key="i" :is="s" />
    </div>
    <slot />
  </modal>
</template>
```

But it's pretty unreadable and not user-friendly. A probably better alternative is using an intermediate functional component that renders the array of vnodes:

```vue
<script setup>
// ...
const RenderSlot = ({ slot }) => slot
</script>

<template>
  <modal>
    <div v-if="HeaderSlot" class="modal-header">
      <RenderSlot :slot="HeaderSlot" />
    </div>
    <slot />
  </modal>
</template>
```

You can find a working example [here](https://sfc.vuejs.org/#eNq1Vm2P0zgQ/iuz2RPNoiZWq5PuLtcuQkjoTkKAWARf8iVNnDZLYlu2U0BV/zvjt2waVrC7LF9ae2b8eGb8+HEO0XMh0n1PoyxaqVI2QoOiuheXOWs6waWGA0hawxFqyTuYYehscL3gnfD2lJiJQTp1L078i+8DlicBSx+Qs5IzpaFTW1ibBOLZf7RtOXzksq3OZhc5WxGXL2aKE0070Raa4gxgtVtcHg528fG4IjizVrOBHeE4xMN5zbk35nrVMNFr2Ccdr2i7ziOEyKOwhpxuYhO+N3JTI2xdtIrm0cM2AnIzdpU9Yj2jVt0V+6EVjbca17R87Joc4L2wH1rTeCtb00lUNI8c85OuEOm14gyv3cGsyL1D5VEG1mJseBfMPI92WguVEaLq0lyQa5VyuSU4SmXPdNPRlKou2Uj+WVGJwHk0H2EQNO6pTCRlFZVU/ghzEvodroE95uyIpaiWa7RhCYNSNOrDa2zV3NzsjjI9h5ey2JrRREByRogzhDS2jd71m7TknanrWpGSS0o2Ld+QrmgYEUX5qdhSRXzFifUrWZIdbQWV6LA5X5mstDp/9fefyat//soZ/WKTq3tW6oYz2FJtYl5wpjGveM8wYXXhmt7UEJ8Fi0QdlAxY37bG56fOmyre0bjcNW0F68twYna574FzXgwwWuJhDlHWmeqvgsJ6vQ7tGoId84ZoNwKYrBp6++RJiDibVOdW2F9sD6qmCbptm0mWRx/6zNfrZplvBhJg6KvT6XdD71Gu4wMYbiCI6Y0ZWuqHB+JHTw0eP15AWs2hV9TAqdtfnsPkGOfjDMKKlHiK4jq8Cv5JMTaFWYYNYizVeVAJfAUhjTi2JUyaahFSjH6WYqm60S3NYPbe/M+w6AsERHY7zJcDpoPye6T7ojWdHj1ikyesavZeaKzmx1euY1emY/5wTMygVQ4X1apsC6XQYvsexAqDz5JkFbIhl0ly4xm1LjOrTuCcMBt9CxmF0Ym04VTpr61LPrXH7y+FKKqqYdsMFkvx5V9n2+Bl3kresyqD86KurBlZhe3wIJ4v9oPhhjBjqtyTAJ6uFa2LvvXJmWPmDFerQXRvUKyCHu1viDWUGCIDW/BcvWW4RRO+6F2j0j9+xhqH4TZ0WwKQpwOBbtnGMcqih/NyGPCUBJ2+O8UWEL8RViGfv/3/wQwbWgpZY2J8+pZIj805zxH7zXg7R36bnPw6n6z0QSwkF2qOCThZQsn0Cx9dkiyf7ihLI5INpPMhnp8Wy6N4E3LtvpxbQmz+uGp+mXi/T9qi4zcxx1Jf) with all API styles (Options API, Composition API, script setup).

A big developer experience improvement could still be made for problem 2, by directly being able to render an array of vnodes in a template, thus removing the need for an intermediate component:

```vue
<template>
  <modal>
    <div v-if="HeaderSlot" class="modal-header">
      <HeaderSlot /> <!-- Tag? -->
    </div>
    <slot />
  </modal>
</template>
```

Note: I like the tag syntax but it implies being able to put attributes on it which wouldn't work. Maybe the compiler can raise warnings about this?

# Detailed design

`getFilledSlot` is a function that takes an array of vnodes and returns it if it's not empty (i.e. contains valid vnodes), `null` otherwise to facilitate conditionals.
It could be implemented like this:
```ts
export function getFilledSlot (vnodes: VNodeArrayChildren | null) {
  return vnodes ? ensureValidVNode(vnodes) : null
}
```

Usage:

```js
const SomeSlot = computed(() => getFilledSlot(slots.some?.({ prop: 'hello' })))
const isSomeSlotFilled = computed(() => SomeSlot.value != null)
```

Note: the returned result from the slot function can be `null` hence the additional conditional in `getFilledSlot`.

---

Allow putting a variable containing an array of vnodes to be used as template tag:

```vue
<template>
  <!-- Array of vnodes from the slot -->
  <SomeSlot />
</template>
```

Usage in Option API could be improved if we could resolve template tag using the computed properties:

```vue
<script>
export default {
  computed: {
    HeaderSlot () {
      return getFilledSlot(this.$slots.header?.({ title: this.title }))
    }
  }
}
<script>

<template>
  <HeaderSlot />
</template>
```

# Drawbacks

`getFilledSlot`: This could be implemented in user space but with hacky solutions involving copying an internal function.

# Adoption strategy

This is an additive API.

# Unresolved questions

- Naming of `getFilledSlot`?
- Choose a solution to render vnodes in a template?

**Comment by stafyniaksacha** [comment]
This would be really helpful. 
About the name, maybe `resolveSlot`?

Note, this would be also really useful to have a signature with a string as first parameter, for the slot name, and slot parameters optional second argument

```ts
getFilledSlot (slot: string, slotProps: any) 
getFilledSlot (vnodes: VNodeArrayChildren | null)
```

**Comment by Akryum** [comment]
The userland implementation I shared [has an issue](https://sfc.vuejs.org/#eNq9V21z4zQQ/itbFy7uXWw1KTOASdopxxzczA0wlJcPmA9urCQutuSR5HCdTP47qzfHcZPScNfrh0Zarfbl0aPVeh1c13W8amiQBBM5E0WtQFLV1JcpK6qaCwVrEHQOG5gLXsEAVQft0mte1U4eEz3RlnaXRzvro4cK4x2F8UOFix2FC6eQshlnUkElFzDVEYaDH2hZcviDizI/GZx5hTxTmdM4RyEm95YpKlZZGYZnML2EdcrAaMUoa+irVynbDGF0fq7VJ8SCgnDgRNGqLjNFcQYwWY4u12sTwGYzITgzUh2kGeHY68PpnHMnTNWkYHWjYBVVPKflNA3QRBq4ZQC0aWLebJwVsuvWwHC0r2KOjuZZKWkaHHQ9kXXGeso6RxcPgoHr3uHjYQEx4yiKWpEF6Flh6ZzBU719GmC6ge2BZvz80FgXR3n7NNB0A+tBcxJF8O1v31/Bd5xKNlDQ1GiWQi14TUV5j4rbzRcPsnJBmUg+hzFMp1M4x1ROs5kqsDhsc+hcpIPA+L8+2nsSs6Lu7eyv69S6gVLEbRsXXDOullSAE3S2+4xtMcTxjulgGNjKGVVZHd9JzrCumwKXugWZBokteVqGtVTP02CpVC0TQuR8pgvsnYy5WBAcxaJhqqhoTGUV3Qr+j6QCDafBsGODoHBFRSQoy6mg4jGbPdUHdrVZRHaDqciSK5RhCu1TVMjff0QaDvXLUFGmhvBGZAs96r1QKSPECnwYi0Itm9t4xiud150kMy4ouS35LamygpE6m/2dLagkLuPIrEsxI0taIt1wwcR8o6NS8vTdV19E777+MmX0vQlu3jBzWrCgSuu85vjOMBWuGAYszyzoxRzCEy8R+NAKBqwpS73mpnY1lryi4WxZlHn7SLntDgO7eNaaUQIPs9Uyi7G6r6mhvYOrVbYXtdX2pO3tarF98cJrnPSyszvMf4QHX0yttM9NL8qNU71y+dpZ4sBAApgDdNDal/yXFn58z8M1aHqgHQ2PHuoth/UNgCyraAKD7cLAsFiXE5nAnwNtZvCXkdmzBmwRHPQ+/mWBZ2PcYRKoiqEiVX3z81gbhczDukrzITSSau9yf1e17jFo2M3E74iJux24D2+ha3a0TGK23kHYtkFYxxwSPgzX/vTO01iIUfsqRohVoUqN2K/6d4Bgn6FBPBdr801r05pyPmwftdM79TqnvFi5OmhahfDGInajEXO80Drtq2LtYuWelZmUKDEH0HmDdDX10RBfI81KB7pE79oxZ18bXU19RH60U1VxKtV9aYM3h+9JUWd5XrBFAqNx/f4bK7vFOrIQvGF5ghV9nhsxsgThcEYcX0wzvCVMlypHEsDRPqfzrCldcPqYOcPdyGwX7daKobhmr6aO1dWUaDU9W7b0by9Ajy/mPnz2X6yxNqxD6xKAvGwJtMeNZZSx7s/L2oCXxD8RT6fYCMKfalOcr39++78Z1kIKSaF1XPiGSB+bc44j5ntoP0eerZx8OJ9M6YPQ1NUhBmDLEpZqt/GjlyTDpyeWpQ7JWtI5FcdPY8tZcSLT6x3HuTGE+ofL4oOJ93ylzdHMfFUffrsOfecf/DrGj4gtGJ2hK/gXST84/R2hE9EvNObh+t4WB7S306Lj/HCLjmgd05lrv4cACjb/Am6WjYo=)
In some cases the vnodes array isn't properly updated. :(

**Reply by Akryum** [reply]
Looks like it's an issue in core: https://github.com/vuejs/core/issues/6202

**Comment by sylvainpolletvillard** [comment]
Detecting slots with empty content is one thing, but I think the need goes beyond that. I recently ran into a problem where my template had to adapt according to the number of elements contained in my slot. Filtering the VNodes was quite cumbersome as you explain, especially if you are not familiar with the VNode API, so instead I [solved my issue with a watcher and DOM queries](https://sfc.vuejs.org/#eNp9U02PmzAQ/SsjLku0xGSvKWHVrirtSu2peys9eGGSkIJt2YZ0FfHfO2MgpanUA7I9n2/ePC7RR2NE32G0jTJX2tp4cOg7kxeqbo22Hp50a2BvdQtFJFJ+cXwRfbhGXMDiHoYx6I6cd4UqVKmV81DqTnm0sOOYeLNiDzV4YWMvmzhewS6fowRZOry/T+Bhs+HYLB0xERp6eGxNIz3SCyA7PoQT4FmfwWs4S18eYa8tuEZzY6qo6DxKdUCQqoID0nBLn3SESpa+7hF6aWv51iA8hurpVD7jgadGmTNS5ZfLhHYYCB5bxoQpMEuvMKMkGhlat9KIk9OKWL5wdDE5XBFtIVjYFmjd0uXovXHbNHX7krk+OaHtIaWbsNS5blGga9dvVp8dkXaiKsmiRkrGHu3aoqrQov1fzZvQf+py2aFQA40yb56V8vcmZn6+EbfbkSWoq10Rte9sKqI8C7Sn+UQYAQ2UjYmZgX5Ne6OERh+gVkAH9c7JY5H4pidzzY8paUFyUMaNcgGWykyuS05o863pPFbJKJcEOoeM0SWg8Jd/rcufN0IesY5i5ikcSXlOikmiS9/TpKvdtc+k75AoKtzLrvHxapHGo4afY0QYf/8RnPwFhPGiLsOfGuSzajhfmM4d40qXXUtOQTL/3CBfP72/VPGfNaweRa0U2ufXr19Cl2H5i0XDb7laYmg=). But if felt wrong and definitely non-intuitive.

May I suggest `getFilledSlot` to return the rendered Nodes instead of the VNodes ? This would make easier to write computed based on various conditions on slot content, and not only if it's empty or not.

**Reply by thedamon** [reply]
This is a good point. DOM queries won't work with SSR either so can introduce jank. It would be nice to be able to interrogate the number of "real" children in a slot (though when one child is a component this becomes convoluted, but that may be acceptable)

**Comment by Akryum** [comment]
Other proposal: update `useSlots().foo()` to return a falsy value if there are no valid vnodes (it's "empty")?
Would be a breaking change though

**Reply by Akryum** [reply]
Or maybe a new composable `useFilledSlots()` or `useValidSlots()`?

```vue
<script setup>
import { useValidSlots, computed } from 'vue'

const slots = useValidSlots()
const TitleSlot = computed(() => slots.title?.())
</script>

<template>
  <div v-if="TitleSlot" style="padding: 12px;">
    <TitleSlot />
  </div>
</template>
```

**Reply by domnantas** [reply]
`useSlots().foo()` returning `null` when there are no valid vnodes seems like the most intuitive approach. Similar to how Vue 2 handles empty slots.

Disagree on the additional composable, it would cause confusion when to use `useSlots` and `useValidSlots`

**Reply by thedamon** [reply]
I agree that the ideal usecase is a vnode tree per slot that excludes 'empty' vnodes. I agree that the "best" solution is to change how useSlots works.. however it's true that would be a breaking change and could cause bugs. (maybe much more likely to fix bugs because i think most times you check for slots.slotName truthiness you expect it to be false when it's empty and in Vue 2 that was more likely the case, but it would change behaviour nonetheless).

I feel though, that the return without the empty vnodes would be much more useful overall. And maybe more useful anytime you're working with vnodes, not just with slot content.

**Comment by posva** [comment]
Worth adding as a reference issue:

- https://github.com/vuejs/core/issues/4733

**Comment by posva** [comment]
Another idea of API could be one that only requires changes to the template instead of having to rely on `useSlots()` in the `<script>`:

```vue
<template>
  <!-- renders only if slot is not empty -->
  <slot-filtered name="foo" v-bind="stuff">
    <div class="wrapper">
      <!-- automatically inherits everything from the <slot-filtered> parent -->
      <slot></slot>
    </div>
  </slot-filtered>
</template>
```

This would generate a different compiled code than the existing one that also renders the slot just once (maybe a `renderSlotWithContent()` function) and is tree-shakeable.

Everything goes on the `slot-filtered` instead of going on `slot`. If the slot has no content (based on the internal `ensureValidVNode()` function), it will skip rendering te `div.wrapper` element, otherwise render the slot content as children of the div

**Reply by Akryum** [reply]
Or we could keep `<slot>` and have a way to decide where to put the slot content:

```vue
<template>
  <slot name="description" v-bind="props">
    <div class="padding">
      <slot-content />
    </div>
    
    <template #fallback>
      Default content
    </template>
  </slot>
</template>
```

**Reply by posva** [reply]
yes! Maybe having the content within `<slot>` makes more sense from the compiler perspective

**Reply by thedamon** [reply]
Oh this is interesting. it solves the essential problem without mucking around with new functions and APIs and just relies on the already existent show 'default' behaviour. I do worry a little because the variety of options for declaring slots in a template is already a bit difficult to reason about.

**Comment by thedamon** [comment]
Strange and different idea to solve the problem:

An `app.config` setting to automatically filter out empty `vnodes` from the return of things like `useSlots` and `children` when called in component code.

I think most developers intuitively assume that `vnodes.length = 1` means there is content in there, and the logic that Vue itself uses to decide whether to display default slot content could be considered the official method of determining that. That could also be a way of making a 'breaking' change much easier to introduce as the old behaviour would be a simple toggle.

The idea of modifying `useSlots` is nice because it elegantly makes what often is confusing go away without needing to think about the difference and decide whether to use an obscure utility function to get the behaviour that seems like what it should actually be.

Scrapping all that, another way that might be the easiest to introduce, would be exporting `ensureValidVNode` from vue (or maybe a pluralized version, or option that takes an Array). I dislike the term 'valid' outside of the usecase in vue itself, but it is difficult to decide on the right terminology here. As a utility function it probably could be used a lot in userland, even as a best practice where it wouldn't be today.

If I were to solve this in my own codebase today I'd make a composable called `useSlotContent(slotName)` and have it return slot content with "invalid" vnodes filtered out, but `null` rather than an empty array if there are none left to allow simple Boolean checks. That solves @sylvainpolletvillard's issue about wanting to check length of slot content. (again the idea of an 'alternate' useValidSlots solves this similarly, but in general an additional method that's basically the same as an existing one does seem unfortunate)

**Reply by Tofandel** [reply]
A config option would make it a nightmare for libraries because you can't assume the value of the config

**Comment by MLMdlb** [comment]
Is this rfc still in discussion ? Feels like not being able to catch an empty slot is an issue.
